source/particle/world.cc

```diff
@@ -0,0 +1,889 @@
+                                                                           std_cxx11::_1));
+    }
+
+    template <>
+    World<3>::World()
+    {
+      data_offset = numbers::invalid_unsigned_int;
+      integrator = NULL;
+      aspect::internals::SimulatorSignals::register_connector_function_3d (std_cxx11::bind(&World<3>::connect_to_signals,
+                                                                           std_cxx11::ref(*this),
+                                                                           std_cxx11::_1));
+    }
+
+    template <int dim>
+    World<dim>::~World()
+    {}
+
+    template <int dim>
+    void
+    World<dim>::initialize(std_cxx11::shared_ptr<Generator::Interface<dim> > particle_generator,
+                           std_cxx11::shared_ptr<Integrator::Interface<dim> > particle_integrator,
+                           std_cxx11::shared_ptr<Interpolator::Interface<dim> > property_interpolator,
+                           std_cxx11::shared_ptr<Property::Manager<dim> > manager,
+                           const ParticleLoadBalancing &load_balancing,
+                           const unsigned int max_part_per_cell,
+                           const unsigned int weight)
+    {
+      generator = particle_generator;
+      integrator = particle_integrator;
+      interpolator = property_interpolator;
+      property_manager = manager;
+      particle_load_balancing = load_balancing;
+      max_particles_per_cell = max_part_per_cell;
+      tracer_weight = weight;
+
+#if DEAL_II_VERSION_GTE(8,4,0)
+      if (particle_load_balancing == repartition)
+        this->get_triangulation().signals.cell_weight.connect(std_cxx11::bind(&aspect::Particle::World<dim>::cell_weight,
+                                                                              std_cxx11::ref(*this),
+                                                                              std_cxx11::_1,
+                                                                              std_cxx11::_2));
+#else
+      AssertThrow(particle_load_balancing != repartition,
+                  ExcMessage("You tried to select the load balancing strategy 'repartition', "
+                             "which is only available for deal.II 8.4 and newer but the installed version "
+                             "seems to be older. Please update your deal.II or choose a different strategy."));
+#endif
+    }
+
+    template <int dim>
+    const Property::Manager<dim> &
+    World<dim>::get_property_manager() const
+    {
+      return *property_manager;
+    }
+
+    template <int dim>
+    std::multimap<types::LevelInd, Particle<dim> > &
+    World<dim>::get_particles()
+    {
+      return particles;
+    }
+
+    template <int dim>
+    const std::multimap<types::LevelInd, Particle<dim> > &
+    World<dim>::get_particles() const
+    {
+      return particles;
+    }
+
+    template <int dim>
+    types::particle_index
+    World<dim>::get_global_particle_count() const
```

Comment:
The implementation of the function is actually a bit dangerous because it requires communication. You can not do certain things with such functions because they may deadlock if called a different number of times from different processors. Is there a way to store this information somewhere once at the beginning of each cycle/time step/... and then just read it from this variable?